services/worker-manager/src/providers/azure.js

```diff
@@ -0,0 +1,379 @@
+      cfgs.push(cfg);
+      toSpawn -= cfg.capacityPerInstance;
+    }
+
+    await Promise.all(cfgs.map(async cfg => {
+      // This must be unique to currently existing instances and match [a-z]([-a-z0-9]*[a-z0-9])?
+      // The lost entropy from downcasing, etc should be ok due to the fact that
+      // only running instances need not be identical. We do not use this name to identify
+      // workers in taskcluster.
+      const resourceGroupName = this.providerConfig.resourceGroupName;
+      const poolName = workerPoolId.replace(/[\/_]/g, '-').slice(0, 38);
+      const virtualMachineName = `vm-${poolName}-${slugid.nice().replace(/_/g, '-').toLowerCase()}`.slice(0, 38);
+      const ipAddressName = `pip-${slugid.nice().replace(/[\/\_]/g, '-').toLowerCase()}`.slice(0, 24);
+      const networkInterfaceName = `nic-${slugid.nice().replace(/[\/\_]/g, '-').toLowerCase()}`.slice(0, 24);
+      const diskName = `disk-${slugid.nice().replace(/[\/\_]/g, '-').toLowerCase()}`.slice(0, 24);
+      let ipAddress, networkInterface, virtualMachine;
+
+      let providerData = {
+        location: cfg.location,
+        resourceGroupName: this.providerConfig.resourceGroupName,
+        vm: {
+          name: virtualMachineName,
+          location: cfg.location,
+        },
+        ip: {
+          name: ipAddressName,
+          location: cfg.location,
+        },
+        nic: {
+          name: networkInterfaceName,
+          location: cfg.location,
+        },
+        disk: {
+          name: diskName,
+          location: cfg.location,
+        },
+      };
+
+      try {
+        // create NIC, public IP, and VM
+        ipAddress = await this._enqueue('query', () => this.networkClient.publicIPAddresses.createOrUpdate(resourceGroupName, ipAddressName, {
+          location: cfg.location,
+          publicIPAllocationMethod: 'Dynamic',
+        }));
+
+        networkInterface = await this._enqueue('query', () => this.networkClient.networkInterfaces.createOrUpdate(resourceGroupName, networkInterfaceName, {
+          location: cfg.location,
+          ipConfigurations: [
+            {
+              name: ipAddressName,
+              privateIPAllocationMethod: 'Dynamic',
+              subnet: {
+                id: cfg.subnetId,
+              },
+              publicIPAddress: {
+                id: ipAddress.id,
+              },
+            },
+          ],
+        }));
+
+        const customData = Buffer.from(JSON.stringify({
+          workerPoolId,
+          providerId: this.providerId,
+          workerGroup: this.providerId,
+          rootUrl: this.rootUrl,
+          workerConfig: cfg.workerConfig || {},
+        })).toString('base64');
+
+        virtualMachine = await this._enqueue('query', () => this.computeClient.virtualMachines.createOrUpdate(resourceGroupName, virtualMachineName, {
+          ...cfg,
+          osProfile: {
+            adminUsername: 'worker',
```

Comment:
I think @grenade was thinking this should be `Administrator` or something like that?

Reply:
Let's move it below the `...cfg.osProfile` as well so that people can't overwrite it.